votor/src/consensus_pool_service.rs

```diff
@@ -412,3 +412,269 @@ impl ConsensusPoolService {
+                if matches!(certificate.certificate, Certificate::Skip(3)) {
+                    return true;
+                }
+            }
+            false
+        });
+        setup_result.exit.store(true, Ordering::Relaxed);
+        consensus_pool_service.join().unwrap();
+    }
+
+    #[test]
+    fn test_produce_block_event() {
+        let setup_result = setup();
+        // Find when is the next leader slot for me (validator 0)
+        let my_pubkey = setup_result.validator_keypairs[0].node_keypair.pubkey();
+        let next_leader_slot = setup_result
+            .leader_schedule_cache
+            .next_leader_slot(&my_pubkey, 0, &setup_result.root_bank.load(), None, 1000000)
+            .expect("Should find a leader slot");
+        // Send skip certificates for all slots up to the next leader slot
+        for slot in 1..next_leader_slot.0 {
+            let skip_certificate = CertificateMessage {
+                certificate: Certificate::Skip(slot),
+                signature: BLSSignature::default(),
+                bitmap: vec![],
+            };
+            setup_result
+                .consensus_message_sender
+                .send(ConsensusMessage::Certificate(skip_certificate))
+                .unwrap();
+        }
+        // Verify that we received a produce block event
+        wait_for_event(&setup_result.event_receiver, |event| {
+            if let VotorEvent::ProduceWindow(LeaderWindowInfo { start_slot, .. }) = event {
+                assert_eq!(*start_slot, next_leader_slot.0);
+                true
+            } else {
+                false
+            }
+        });
+        setup_result.exit.store(true, Ordering::Relaxed);
+        setup_result.consensus_pool_service.join().unwrap();
+    }
+
+    #[test]
+    fn test_standstill() {
+        let setup_result = setup();
+        // Do nothing for 10.1 seconds
+        thread::sleep(Duration::from_millis(10_100));
+        // Verify that we received a standstill event
+        wait_for_event(&setup_result.event_receiver, |event| {
+            if let VotorEvent::Standstill(slot) = event {
+                assert_eq!(*slot, 0);
+                true
+            } else {
+                false
+            }
+        });
+        setup_result.exit.store(true, Ordering::Relaxed);
+        setup_result.consensus_pool_service.join().unwrap();
+    }
+
+    #[test]
+    fn test_channel_disconnection() {
+        let setup_result = setup();
+        drop(setup_result.consensus_message_sender);
+        // Verify that the consensus pool service exits within 5 seconds
+        let start = Instant::now();
+        while !setup_result.exit.load(Ordering::Relaxed) && start.elapsed() < Duration::from_secs(5)
+        {
+            thread::sleep(Duration::from_millis(100));
+        }
+        assert!(setup_result.exit.load(Ordering::Relaxed));
+        setup_result.consensus_pool_service.join().unwrap();
```

Comment:
can we make a generic test that takes a receiver and tests that this happens for all the  receiver channels as well, `bls_receiver`, `event_receiver`, `commitment_receiver`.